### PR TITLE
Don't break single argument destructuring arguments

### LINKED ISSUE #3016: Prettier with long list of optional named parameters (using destructuring) not so pretty
I have this class:

```javascript
class Model {
  /**
   * @param id {String}
   * @param createdAt {Date}
   * @param updatedAt {Date}
   * @param endedAt {Date}
   */
  constructor({
      id = uuid(),
      createdAt = new Date(),
      updatedAt = new Date(),
      endedAt = null,
  } = {}) {
    this.id = id;
    this.createdAt = createdAt;
    this.updatedAt = updatedAt;
    this.endedAt = endedAt;
  }
}
```

Prettier makes this from it:

```javascript
class Model {
  /**
   * @param id {String}
   * @param createdAt {Date}
   * @param updatedAt {Date}
   * @param endedAt {Date}
   */
  constructor(
    {
      id = uuid(),
      createdAt = new Date(),
      updatedAt = new Date(),
      endedAt = null,
    } = {},
  ) {
    this.id = id;
    this.createdAt = createdAt;
    this.updatedAt = updatedAt;
    this.endedAt = endedAt;
  }
}
```

Which feels a bit weird to me. If I remove the default value of `{}`, it becomes much prettier:

```javascript
class Model {
  /**
   * @param id {String}
   * @param createdAt {Date}
   * @param updatedAt {Date}
   * @param endedAt {Date}
   */
  constructor({
    id = uuid(),
    createdAt = new Date(),
    updatedAt = new Date(),
    endedAt = null,
  }) {
    this.id = id;
    this.createdAt = createdAt;
    this.updatedAt = updatedAt;
    this.endedAt = endedAt;
  }
}
```

But now I need to always explicitly give the first argument, as `new Model()` gives now an error (wished Javascript behaved differently). I have to write `new Model({})` in that case, if I want the default values.

### LINKED ISSUE #3046: Parentheses should hug when destructuring array params
**Prettier 1.7.4**
[Playground link](https://prettier.io/playground/#%7B%22content%22%3A%22promise.then((%5Cn%20%20%5Baaaaaa%2C%20bbbbbb%2C%20ccccccc%2C%20dddddddd%2C%20eeeeeee%2C%20ffffffff%2C%20gggggggg%2C%20hhhhhhh%5D%5Cn)%20%3D%3E%20%7B%5Cn%20%20console.log(eeeee)%3B%5Cn%7D)%3B%5Cn%22%2C%22options%22%3A%7B%22ast%22%3Afalse%2C%22bracketSpacing%22%3Atrue%2C%22doc%22%3Afalse%2C%22jsxBracketSameLine%22%3Afalse%2C%22output2%22%3Afalse%2C%22parser%22%3A%22babylon%22%2C%22printWidth%22%3A80%2C%22semi%22%3Atrue%2C%22singleQuote%22%3Afalse%2C%22tabWidth%22%3A2%2C%22trailingComma%22%3A%22none%22%2C%22useTabs%22%3Afalse%7D%7D)

**Input:**
```jsx
promise.then((
  [aaaaaa, bbbbbb, ccccccc, dddddddd, eeeeeee, ffffffff, gggggggg, hhhhhhh]
) => {
  console.log(eeeee);
});

```

**Output:**
```jsx
promise.then(
  (
    [aaaaaa, bbbbbb, ccccccc, dddddddd, eeeeeee, ffffffff, gggggggg, hhhhhhh]
  ) => {
    console.log(eeeee);
  }
);

```

**Expected behavior:**
Input preserved, probably as long as it's the only parameter

### PR DESCRIPTION
**Pull request by duailibe**
Fixes #3016
Fixes #3046

### CODE REVIEW

File: tests/flow/destructuring/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -165,9 +165,9 @@ obj_prop_var({}); // ok
 obj_prop_var({ p: {} }); // ok
 obj_prop_var({ p: { q: null } }); // ok, provides add'l lower bound
 
-function obj_rest(
-  { p: { q, ...o } = { q: 0, r: 0 } } = { p: { q: 0, r: "" } }
-) {
+function obj_rest({ p: { q, ...o } = { q: 0, r: 0 } } = {
+  p: { q: 0, r: "" }
+}) {
```
Comment (duailibe) [inline tests/flow/destructuring/__snapshots__/jsfmt.spec.js.snap]: this is weird but I think it's an edge case.. we could check if the right side is not another `AssignmentPattern`

File: tests/flow/destructuring/__snapshots__/jsfmt.spec.js.snap [hunk 2]
```diff
@@ -179,15 +179,13 @@ obj_rest({}); // ok
...
-    p: 0 // error: number ~> string
-  }
-) {
+function obj_prop_annot({
+  p = true // error: boolean ~> string
+}: {
+  p: string
+} = {
+  p: 0 // error: number ~> string
+}) {
```
Comment (duailibe) [inline tests/flow/destructuring/__snapshots__/jsfmt.spec.js.snap]: not sure if this is better or worse

**Comment by vjeux** [comment]
I'm not sure about this one, the hugging behavior is really nice when there's just a single `{}` but if you start adding default values then it feels a bit clunky. I'm not opposed to it but not super thrilled either.

**Comment by duailibe** [comment]
I have no strong opinions on this one either.. I do think that hugging on `= {}` and `= default` is nice, not sure about it when it's a big default object

**Comment by vjeux** [comment]
Can you use a whitelist approach for those instead of allowing everything?

**Comment by duailibe** [comment]
ofc :-)

That solves the one case was concerned about

File: tests/function_single_destructuring/__snapshots__/jsfmt.spec.js.snap [hunk 3]
```diff
@@ -93,6 +205,19 @@ function StatelessFunctionalComponent({
...
+function StatelessFunctionalComponent({
+  isActive = true,
+  onFiltersUpdated = () => null,
+  onSelect = () => null,
+  onSubmitAndDeselect = () => null,
+  onCancel = () => null,
+  searchFilters = null,
+  title = "",
+  items = []
+} = {}) {
```
Comment (azz) [inline tests/function_single_destructuring/__snapshots__/jsfmt.spec.js.snap]: Could you add a test when there's multiple props in the rhs object?

**Pull request merged**